mempool/clist_mempool.go

```diff
@@ -210,10 +221,12 @@ func (mem *CListMempool) TxsWaitChan() <-chan struct{} {
 // cb: A callback from the CheckTx command.
 //     It gets called from another goroutine.
 // CONTRACT: Either cb will get called, or err returned.
+//
+// Safe for concurrent use by multiple goroutines.
 func (mem *CListMempool) CheckTx(tx types.Tx, cb func(*abci.Response), txInfo TxInfo) (err error) {
-	mem.proxyMtx.Lock()
+	mem.updateMtx.RLock()
```

Comment:
I'm not sure if this is entirely safe. Below, we're first checking that the mempool isn't full, then later sending the tx to the ABCI app and recording its size. However, since multiple txs can now execute this concurrently, it is possible for multiple txs to pass the check (because they individually did not exceed the limit), but in sum they will exceed the limit. However, this may already be a problem, since we're calling CheckTx async on the ABCI app. Not familiar enough with the code to say with any certainty.

Reply:
Good catch! It's already an issue. We just haven't seen it because, well, most of folks use localClient which is completely synchronous!, and because we have a `mem.proxyMtx.Lock()` - no concurrency at all 😞

I personally think it's fine for mempool sometimes to be bigger IF we document this explicitly. Kinda like a trade-off for better concurrency.

Reply:
> I personally think it's fine for mempool sometimes to be bigger IF we document this explicitly. Kinda like a trade-off for better concurrency.

Yes, it may be fine. What I'm most worried about is whether it might be a DoS vector (e.g. someone sending enough maliciously crafted txs to OOM the node), but I don't know how easy that might be to trigger. As you say, it's a trade-off.

I think it would be fixed by simply re-checking that the mempool size isn't exceeded when we're actually adding the tx, after the ABCI app has checked it. That is already serialized due to the clist mutex, and the operation is cheap, while still allowing concurrent ABCI tx checking.

Reply:
> I think it would be fixed by simply re-checking that the mempool size isn't exceeded when we're actually adding the tx

Yep, let's do that.